libfreerdp/core/nla.c

```diff
@@ -1634,8 +1600,7 @@ static int nla_server_authenticate(rdpNla* nla)
 				}
 			}
 
-			if (!nla_complete_auth(nla, &outputBufferDesc))
-				goto fail;
+			hash_set = TRUE;
 		}
```

Comment:
Ok for not setting multiple time the hash callback and data, but why removing the `nla_complete_auth()` call. Does server-side still work after your changes ?

Reply:
So it seems that the return value `SEC_I_COMPLETE_NEEDED` was being used to signal that certain information (SamFile / HashCallback) was needed to continue setting up the context. They don't need to be set at any particular stage; but it needs to be after at least one call to `AcceptSecurityContext()` so that there is a context to set it in (in fact, fwiw, I think it would be better to have them passed into `AquireCredentialHandle()` and stored in the credential). Anyway, it's difficult to use `COMPLETE_NEEDED` in this way with spnego because `AcceptSecurityContext()` needs the context set up to use the signature functions before returning the last call

Reply:
After looking this over I realized this won't work when NTLM is not used directly, because there still won't be an ntlm context after the first call. I still think the best option is to store the information in the credential

Reply:
Should we move the callback function's definition to sspi.h, do you think? Also, do you think we should add elements to `SSPI_CREDENTIALS` or define a new `NTLM_CREDENTIALS` structure?